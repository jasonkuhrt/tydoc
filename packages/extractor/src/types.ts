export * as Doc from './extractor/doc'
