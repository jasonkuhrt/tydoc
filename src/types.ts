export * as Doc from './lib/extractor/doc'
