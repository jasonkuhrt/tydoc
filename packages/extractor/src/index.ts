export * from './api/extractor/extract'
export { fromPublished } from './api/extractor/fromPublished'
export { Options as RenderMarkdownOptions, render as renderMarkdown } from './api/renderers/markdown'
