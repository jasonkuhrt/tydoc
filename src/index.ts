import * as tsm from 'ts-morph'
import { inspect } from 'util'

export function extractDocsAndTypesFromModuleAtPath(filePath: string) {
  const project = new tsm.Project({ addFilesFromTsConfig: true })

  project.addSourceFileAtPathIfExists(filePath)

  const source = project.getSourceFile(filePath)

  if (!source) {
    throw new Error(`No file at ${filePath}`)
  }

  return extractDocsAndTypesFromSourceFile(source)
}

function extractDocsAndTypesFromSourceFile(sourceFile: tsm.SourceFile) {
  const exs = sourceFile.getExportedDeclarations()

  const docs = []
  for (const [name, declarations] of exs) {
    docs.push(extractDocsAndTypesFromDeclaration(name, declarations[0]))
  }

  return docs
}

interface JSDocBlock {
  source: string
}
/**
 * It is possible for multiple jsDoc blocks to appear in succession. When
 * source code is authored that way, the jsDoc blocks _after_ the one closest
 * to the code item appear here. For example:
 *
 *    /**
 *     *   foo 2           <-- Goes into additional jsDoc blocks group
 *     *\/
 *    /**
 *     *   foo 1           <-- Considered the primary jsDoc block
 *     *\/
 *     const a = 1
 *
 */
interface JSDocContent {
  /**
   * The jsDoc block closest to the code.
   */
  primary: JSDocBlock
  additional: JSDocBlock[]
}

interface DocBase {
  name: string
  jsDoc: null | JSDocContent
  location: {
    filePath: string
    fileLine: number
  }
}

interface DocFunction extends DocBase {
  kind: 'function'
  signature: {
    parameters: any[]
    return: any
  }
}

interface DocVariable extends DocBase {
  kind: 'variable'
  type: any
}

type DocItem = DocFunction | DocVariable

function extractDocsAndTypesFromDeclaration(
  name: string,
  declaration: tsm.ExportedDeclarations
): DocItem {
  const sourceFile = declaration.getSourceFile()
  const filePath = sourceFile.getFilePath()
  const fileLine = declaration.getStartLineNumber()

  if (declaration instanceof tsm.FunctionDeclaration) {
    return {
      kind: 'function',
      name,
      signature: {
        parameters: declaration
          .getParameters()
          .map(
            p => `name: ${p.getName()} - type: ${extractTypeData(p.getType())}`
          ),
        return: declaration.getReturnType().getText(),
      },
      jsDoc: getJSDocContent(declaration),
      location: {
        filePath,
        fileLine,
      },
    }
  }

  if (declaration instanceof tsm.VariableDeclaration) {
    // A variable declaration is within a variable declaration list is inside a
    // variable statement. JSDoc lives at the variable statement level.
    const ctx = declaration.getParent().getParent()
    const initializer = declaration.getInitializer()
    if (ctx instanceof tsm.VariableStatement) {
      // todo when would this be undefined?
      const type = initializer && initializer.getType()
      return {
        kind: 'variable',
        name,
        type: type && extractTypeData(type),
        jsDoc: getJSDocContent(ctx),
        location: {
          filePath,
          fileLine,
        },
      }
    }
  }

  throw new Error(
    `unknown kind of declaration or declaration scenario ${inspect(
      declaration
    )}`
  )
}

/**
 * Extract JSDoc from node. Tease apart primary block from additional blocks.
 */
function getJSDocContent(node: tsm.JSDocableNode): null | JSDocContent {
  const jsDocs = node.getJsDocs().map(doc => ({ source: doc.getFullText() }))

  if (jsDocs.length === 0) return null
  if (jsDocs.length === 1) return { primary: jsDocs[0], additional: [] }

  return { primary: jsDocs.pop()!, additional: jsDocs }
}

type TypeMetaData = {
  text: string
  properties: {
    name: string
    type: any // todo TypeMetaData
  }[]
}

function extractTypeData(type: tsm.Type<tsm.ts.Type>): TypeMetaData {
  const properties = type.getApparentProperties()

  return {
    text: type.getText(),
    properties: properties.map(p => ({
      name: p.getName(),
      type: extractTypeData(p.getDeclaredType()),
    })),
  }
}

// const docs = extractDocsAndTypesFromModuleAtPath(
//   '/Users/jasonkuhrt/projects/nexus/nexus-future/src/index.ts'
// )

// docs //?+
// inspect(docs, { depth: null }) //?
