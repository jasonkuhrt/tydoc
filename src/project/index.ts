export * from './lib'
export { cFunction } from './lib-2'

interface App {
  server: () => any
  inputObjectType: () => any
}

function createApp(): App {
  return {
    a: 'b',
    c: 'd'
  } as any
}

/**
 * yeaaaah
 */
export const a = createApp()
