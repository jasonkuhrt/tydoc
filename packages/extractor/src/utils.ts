import * as tsm from 'ts-morph'
import { inspect } from 'util'
import { dumpNode, dumpType } from './api/extractor/utils'

export type ArrayOrVarg<T> = T[] | [T[]]

export type MaybeArray<T> = T | T[]

export type Index<T> = Record<string, T>

export type Thunk<T> = () => T

/**
 * Use this to make assertion at end of if-else chain that all members of a
 * union have been accounted for.
 */
export function casesHandled(x: never): never {
  throw new Error(`A case was not handled for value: ${x}`)
}

/**
 * Quick and dirty logging. Not for production.
 */
export function dump(...args: any[]) {
  if (args[0] instanceof tsm.Node) return dumpNode(args[0])
  if (args[0] instanceof tsm.Type) return dumpType(args[0])

  const argsInspected = args.map((a) => inspect(a, { depth: 20 }))
  console.error(...argsInspected)
}

export function indexBy<T extends object>(items: T[], property: keyof T | ((item: T) => string)): Index<T> {
  const index = {} as any
  for (const item of items) {
    if (typeof property === 'function') {
      index[property(item)] = item
    } else {
      index[item[property]] = item
    }
  }
  return index
}

export function arrayify<T>(x: T): T extends Array<any> ? T : T[] {
  return Array.isArray(x) ? x : ([x] as any)
}
