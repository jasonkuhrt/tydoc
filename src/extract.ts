import * as tsm from 'ts-morph'
import { inspect } from 'util'
import { casesHandled } from './utils'

/**
 * The root of documentation data.
 */
export interface Docs {
  terms: DocItem[]
  types: DocItem[]
  hybrids: DocItem[]
  length: number
}

/**
 * Create a new set of docs.
 */
function createDocs(): Docs {
  return {
    terms: [],
    hybrids: [],
    types: [],
    length: 0,
  }
}

/**
 * Add a doc item to docs.
 */
function addDoc(docs: Docs, doc: DocItem): Docs {
  docs[
    doc.languageLevel === 'term'
      ? 'terms'
      : doc.languageLevel === 'type'
      ? 'types'
      : doc.languageLevel === 'hybrid'
      ? 'hybrids'
      : casesHandled(doc.languageLevel)
  ].push(doc)
  docs.length++
  return docs
}

interface JSDocBlock {
  source: string
}

/**
 * It is possible for multiple jsDoc blocks to appear in succession. When
 * source code is authored that way, the jsDoc blocks _after_ the one closest
 * to the code item appear here. For example:
 *
 *    /**
 *     *   foo 2           <-- Goes into additional jsDoc blocks group
 *     *\/
 *    /**
 *     *   foo 1           <-- Considered the primary jsDoc block
 *     *\/
 *     const a = 1
 *
 */
interface JSDocContent {
  /**
   * The jsDoc block closest to the code.
   */
  primary: JSDocBlock
  additional: JSDocBlock[]
}

interface TypeData {
  name: string
}

interface DocBase {
  name: string
  jsDoc: null | JSDocContent
  text: string
  /**
   * This is about if the item is from the JavaScript language or the TypeScript
   * type system. "term" refers to JavaScript values. "type" refers to
   * TypeSript types. "hybrid" refers to constructs span both levels, such as classes.
   */
  languageLevel: 'term' | 'type' | 'hybrid'
  sourceLocation: {
    filePath: string
    fileLine: number
  }
}

export interface DocFunction extends DocBase {
  kind: 'function'
  signature: {
    parameters: { name: string; type: TypeData }[]
    return: TypeData
  }
}

export interface DocVariable extends DocBase {
  kind: 'variable'
  type: TypeData
}

export interface DocTypeAlias extends DocBase {
  kind: 'typeAlias'
}

export type DocItem = DocFunction | DocVariable | DocTypeAlias

/**
 * Extract docs from the module found at the given file path.
 */
export function extractDocsFromModuleAtPath(filePath: string) {
  const project = new tsm.Project({ addFilesFromTsConfig: true })

  project.addSourceFileAtPathIfExists(filePath)

  const source = project.getSourceFile(filePath)

  if (!source) {
    throw new Error(`No file at ${filePath}`)
  }

  return extractDocsFromModule(source)
}

/**
 * Extract docs from the given module.
 */
export function extractDocsFromModule(sourceFile: tsm.SourceFile): Docs {
  const exs = sourceFile.getExportedDeclarations()
  const docs = createDocs()

  for (const [name, declarations] of exs) {
    const doc = extractDocsFromDeclaration(name, declarations[0])
    addDoc(docs, doc)
  }

  return docs
}

/**
 * Extracts docs from the given declaration.
 */
function extractDocsFromDeclaration(
  name: string,
  dec: tsm.ExportedDeclarations
): DocItem {
  if (dec instanceof tsm.FunctionDeclaration) {
    return {
      kind: 'function',
      languageLevel: 'term',
      name,
      text: dec.getText(false),
      signature: {
        parameters: dec.getParameters().map(
          // p => `name: ${p.getName()} - type:
          // ${extractTypeData(p.getType())}`
          p => ({ name: p.getName(), type: getTypeData(p.getType()) })
        ),
        return: getTypeData(dec.getReturnType()),
      },
      jsDoc: getJSDocContent(dec),
      ...extractCommonDocDataFromDeclaration(dec),
    }
  }

  if (dec instanceof tsm.VariableDeclaration) {
    // A variable declaration is within a variable declaration list is inside a
    // variable statement. JSDoc lives at the variable statement level.
    const ctx = dec.getParent().getParent()
    if (ctx instanceof tsm.VariableStatement) {
      // If the variable is pointing to a function we will treat it like as if
      // it were a function declaration.
      const initializer = dec.getInitializer()
      if (
        initializer &&
        (initializer instanceof tsm.ArrowFunction ||
          initializer instanceof tsm.FunctionExpression)
      ) {
        const docs = extractDocsFromFunction(initializer) as DocFunction
        docs.jsDoc = getJSDocContent(ctx)
        docs.name = docs.name || name
        return docs
      }

      // const type = initializer && initializer.getType()
      const typeName = dec.getType().getText()
      return {
        kind: 'variable',
        languageLevel: 'term',
        name,
        // type: type && extractTypeData(type),
        type: {
          name: typeName,
        },
        text: dec.getText(false),
        jsDoc: getJSDocContent(ctx),
        ...extractCommonDocDataFromDeclaration(dec),
      }
    }
  }

  if (dec instanceof tsm.TypeAliasDeclaration) {
    return {
      kind: 'typeAlias',
      languageLevel: 'type',
      name,
      text: dec.getText(false),
      jsDoc: getJSDocContent(dec),
      ...extractCommonDocDataFromDeclaration(dec),
    }
  }

  // todo
  // casesHandled(declaration)

  throw new Error(
    `unknown kind of declaration or declaration scenario\n\n${inspect(dec)}`
  )
}

/**
 * Extract doc data that is common to all doc items from the given declaration.
 */
function extractCommonDocDataFromDeclaration(dec: tsm.ExportedDeclarations) {
  const sourceFile = dec.getSourceFile()
  const filePath = sourceFile.getFilePath()
  const fileLine = dec.getStartLineNumber()
  const commonDocData = {
    sourceLocation: {
      filePath,
      fileLine,
    },
  }
  return commonDocData
}

/**
 * Extract docs from the given Function like node. This does not quite return
 * full doc data because jsDoc is kept on variable declarations and name should
 * be the exported one not the maybe-present explicit function expression name.
 */
function extractDocsFromFunction(
  node: tsm.ArrowFunction | tsm.FunctionExpression
): Omit<DocFunction, 'name' | 'jsDoc'> {
  // todo anything useful we can by revealing explicitly named function expressions?
  // let maybeName = ''
  // if (dec instanceof tsm.FunctionExpression) {
  //   maybeName = dec.getName() || ''
  // }
  return {
    kind: 'function',
    languageLevel: 'term',
    text: node.getText(false),
    signature: {
      parameters: node
        .getParameters()
        .map(p => ({ name: p.getName(), type: getTypeData(p.getType()) })),
      return: getTypeData(node.getReturnType()),
    },
    ...extractCommonDocDataFromDeclaration(node),
  }
}

/**
 * Extract JSDoc from node. Tease apart primary block from additional blocks.
 */
function getJSDocContent(node: tsm.JSDocableNode): null | JSDocContent {
  const jsDocs = node.getJsDocs().map(doc => ({ source: doc.getFullText() }))

  if (jsDocs.length === 0) return null
  if (jsDocs.length === 1) return { primary: jsDocs[0], additional: [] }

  return { primary: jsDocs.pop()!, additional: jsDocs }
}

function getTypeData(type: tsm.Type): TypeData {
  return {
    name: type.getText(),
  }
}

// type TypeMetaData = {
//   text: string
//   properties: {
//     name: string
//     type: any // todo TypeMetaData
//   }[]
// }

// function extractTypeData(type: tsm.Type<tsm.ts.Type>): TypeMetaData {
//   const properties = type.getApparentProperties()
//   return {
//     text: type.getText(),
//     properties: properties.map(p => ({
//       name: p.getName(),
//       type: extractTypeData(p.getDeclaredType()),
//     })),
//   }
// }
