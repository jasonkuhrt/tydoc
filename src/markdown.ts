import { Docs } from './extract'

/**
 * Render docs as Markdown.
 */
export function render(docs: Docs): string {
  let md = ''

  md += '### API\n\n'
  md += docs.terms
    .map(term => {
      let s = `#### ${term.name}\n\n`
      if (term.kind === 'function') {
        s += markdownCodeBlock(term.signature.text, 'ts')
      }
      return s
    })
    .join('\n\n')

  md += '\n\n'
  md += '### Types\n\n'
  md += docs.types
    .map(type => {
      let s = ''
      s += `#### ${type.name}\n\n`
      s += type.properties
        .map(prop => {
          let s = ''
          s += `- ${prop.name} (\`${prop.type.name}\`)\n`
          s += `  ${prop.jsDoc ? prop.jsDoc.primary.source + '\n' : ''}`
          return s
        })
        .join('\n')
      return s
    })
    .join('\n\n')

  md += '\n'
  return md
}

/**
 * Render the given string inside markdown codeblock.
 */
function markdownCodeBlock(s: string, languageType?: string): string {
  return '```' + (languageType || '') + '\n' + s + '\n' + '```'
}
