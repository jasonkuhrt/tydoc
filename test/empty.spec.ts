import '../src'
const ctx = createContext()

it('extracts nothing if the module is empty', () => {
  expect(ctx.given('')).toMatchSnapshot()
})

it('extracts nothing if the module has no exports', () => {
  expect(
    ctx.given(`
      function a() {}
      const b = 2
    `)
  ).toMatchSnapshot()
})
