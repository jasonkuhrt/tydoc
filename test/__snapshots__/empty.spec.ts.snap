// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extracts nothing if the module has no exports 1`] = `
Object {
  "hybrids": Array [],
  "length": 0,
  "terms": Array [],
  "types": Array [],
}
`;

exports[`extracts nothing if the module is empty 1`] = `
Object {
  "hybrids": Array [],
  "length": 0,
  "terms": Array [],
  "types": Array [],
}
`;
